src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs

```diff
@@ -215,73 +185,67 @@ public TrialResult Run()
-                if (cts.Token.IsCancellationRequested)
+                if (_cts.Token.IsCancellationRequested)
                 {
                     break;
                 }
                 var setting = new TrialSettings()
                 {
-                    ExperimentSettings = _settings,
                     TrialId = trialNum++,
+                    Parameter = Parameter.CreateNestedParameter(),
                 };
 
-                setting = pipelineProposer.Propose(setting);
-                setting = hyperParameterProposer.Propose(setting);
-                monitor.ReportRunningTrial(setting);
-                var runner = runnerFactory.CreateTrialRunner();
+                var parameter = tuner.Propose(setting);
+                setting.Parameter = parameter;
+                monitor?.ReportRunningTrial(setting);
+                var runner = serviceProvider.GetService<ITrialRunner>();
+                Contracts.Assert(runner != null, "trial runner can't be null");
 
                 try
                 {
-                    var trialResult = runner.Run(setting, serviceProvider);
-                    monitor.ReportCompletedTrial(trialResult);
-                    hyperParameterProposer.Update(setting, trialResult);
-                    pipelineProposer.Update(setting, trialResult);
+                    var trialResult = runner.Run(setting);
+                    monitor?.ReportCompletedTrial(trialResult);
+                    tuner.Update(trialResult);
 
-                    var error = _settings.IsMaximizeMetric ? 1 - trialResult.Metric : trialResult.Metric;
+                    var error = _settings.IsMaximize ? 1 - trialResult.Metric : trialResult.Metric;
                     if (error < _bestError)
                     {
                         _bestTrialResult = trialResult;
                         _bestError = error;
-                        monitor.ReportBestTrial(trialResult);
+                        monitor?.ReportBestTrial(trialResult);
                     }
                 }
+                catch (OperationCanceledException)
+                {
+                    break;
+                }
                 catch (Exception ex)
                 {
-                    if (cts.Token.IsCancellationRequested)
-                    {
-                        break;
-                    }
-                    else
+                    monitor?.ReportFailTrial(setting, ex);
+
+                    if (!_cts.IsCancellationRequested && _bestTrialResult == null)
                     {
                         // TODO
                         // it's questionable on whether to abort the entire training process
                         // for a single fail trial. We should make it an option and only exit
                         // when error is fatal (like schema mismatch).
-                        monitor.ReportFailTrial(setting, ex);
                         throw;
                     }
+                    else
```

Comment:
>else [](http://example.com/codeflow?start=20&length=4)

else not needed here.

Reply:
That `else` will be hit when you get a training error but has a successful trial result(therefore _bestTrialResult is not null). In which case the current best result will be returned instead.

This is to avoid the case of losing all available trial results when encountering an unfatal error, like OOM or so. The more reliable way of doing that is, of course, detecting if exception from trial is fatal or not and continue training if the exception is not fatal. But in that case we need to cover all unfatal cases which is almost impossible and unnecessary. So as a step back, in order not to loss current training result, `AutoMLExperiment` simply 1) prints out exception and 2) return _currentBestTrial if there's any when encountering any exception. Only when there's no completed trial will `AutoMLExperiment` throws an exception.

Reply:
The if block is throwing any way. so no need to have explicit else.

Reply:
OIC